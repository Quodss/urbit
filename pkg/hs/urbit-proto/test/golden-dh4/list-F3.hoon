=>
|%
++  list
  ^-  $-($ $)
  |=  a=$
  $%  [%nil $~]
      [%con a (list a)]
  ==
--
=/  a  `(list @)`[%nil ~]
=/  b  `(list @)`[%con 1 %con 2 %nil ~]
`(list ^)`b
