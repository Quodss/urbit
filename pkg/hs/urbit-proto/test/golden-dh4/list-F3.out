PARSED

=>  |%
    ++  list
      ^-  $-($ $)
      |=  a=$
      $%  {$nil [$~]}
          {$con [a (list a)]}
      ==
    --
=/  a
  `(list @)`[%nil ~]
=/  b
  `(list @)`[%con 1 %con 2 %nil ~]
`(list ^)`b

SOFT

=>  |%
    ++  list
      ^-  $-($ $)
      |=  a=$
      $:  ?(%7104878 %7237475)
          ^-  $
          ?:  ?=(%"nil" +2)
            $~
          ??  ?=(%"con" +2)
          [a (list a)]
      ==
    --
=+  a=`(list @)`[%nil ~]
=+  b=`(list @)`[%con 1 %con 2 %nil ~]
`(list * *)`b

HARD

=>  |%
    ++  list
      ^-  $?  +7
          $|  [$;($ *) *]
          ++  %  $
          --
      =:  a=$
        ^-  *
        0
      |%
      ++  %
        $:  ?(%7104878 %7237475)
            ^-  $
            ?:  ?=(%"nil" +2)
              $~
            ??  ?=(%"con" +2)
            :-  a
            =+  list
            %.+2(+6 a:+3)
        ==
      --
    --
=+  ^=  a
    ^!  =+  list
        %.+2(+6 @:+3)
    [%nil ~]
=+  ^=  b
    ^!  =+  list
        %.+2(+6 @:+3)
    [%con 1 %con 2 %nil ~]
^!  =+  list
    %.+2(+6 [* *]:+3)
b

ERROR

  trace:  play:
          lvl   0
          sut   0|*
          ----
          code  =>  |%
                    ++  list
                      ^-  $-($ $)
                      |=  a=$
                      $:  ?(%7104878 %7237475)
                          ^-  $
                          ?:  ?=(%"nil" +2)
                            $~
                          ??  ?=(%"con" +2)
                          [a (list a)]
                      ==
                    --
                =+  a=`(list @)`[%nil ~]
                =+  b=`(list @)`[%con 1 %con 2 %nil ~]
                `(list * *)`b


          play:
          lvl   0
          sut   $=  |%
                    ++  list
                      |=  _
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [+14 (list.+15 +14)]
                      ==
                    --
                $|  0|*
                ++  list  $-($ $)
                --

          ----
          code  =+  a=`(list @)`[%nil ~]
                =+  b=`(list @)`[%con 1 %con 2 %nil ~]
                `(list * *)`b


          play:
          lvl   0
          sut   $=  |%
                    ++  list
                      |=  _
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [+14 (list.+15 +14)]
                      ==
                    --
                $|  0|*
                ++  list  $-($ $)
                --

          ----
          code  a=`(list @)`[%nil ~]

          nest:
          have  =>  =>  list
                    .(+6 @)
                %

          need  $

          nest:
          have  %
          need  $

          play:
          lvl   1
          sut   :-  $=  ["nil" 0]
                    ^=  a
                    =>  =>  list
                        .(+6 @)
                    %
                $=  |%
                    ++  list
                      |=  _
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [+14 (list.+15 +14)]
                      ==
                    --
                $|  0|*
                ++  list  $-($ $)
                --

          ----
          code  =+  b=`(list @)`[%con 1 %con 2 %nil ~]
                `(list * *)`b


          play:
          lvl   1
          sut   :-  $=  ["nil" 0]
                    ^=  a
                    =>  =>  list
                        .(+6 @)
                    %
                $=  |%
                    ++  list
                      |=  _
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [+14 (list.+15 +14)]
                      ==
                    --
                $|  0|*
                ++  list  $-($ $)
                --

          ----
          code  b=`(list @)`[%con 1 %con 2 %nil ~]

          nest:
          have  =>  =>  list
                    .(+6 @)
                %

          need  $

          nest:
          have  %
          need  $

          play:
          lvl   2
          sut   :+  $=  ["con" 1 "con" 2 "nil" 0]
                    ^=  b
                    =>  =>  list
                        .(+6 @)
                    %
                  $=  ["nil" 0]
                  ^=  a
                  =>  =>  list
                      .(+6 @)
                  %
                $=  |%
                    ++  list
                      |=  _
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [+14 (list.+15 +14)]
                      ==
                    --
                $|  0|*
                ++  list  $-($ $)
                --

          ----
          code  `(list * *)`b

          work:
          lvl   2
          sut   :+  $=  ["con" 1 "con" 2 "nil" 0]
                    ^=  b
                    =>  =>  list
                        .(+6 @)
                    %
                  $=  ["nil" 0]
                  ^=  a
                  =>  =>  list
                      .(+6 @)
                  %
                $=  |%
                    ++  list
                      |=  _
                      $:  ?(%7104878 %7237475)
                          ?:  ?=(%"nil" +2)
                            $~
                          [+14 (list.+15 +14)]
                      ==
                    --
                $|  0|*
                ++  list  $-($ $)
                --

          ----
          mode  cast
          code  b
          type  =>  =>  list
                    .(+6 [* *])
                %


          cast:
          have  =>  =>  list
                    .(+6 @)
                %

          need  =>  =>  list
                    .(+6 [* *])
                %


          cast:
          have  %
          need  =>  =>  list
                    .(+6 [* *])
                %


          cast:
          have  $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    :-  @
                    =>  =>  list
                        .(+6 @)
                    %
                ==

          need  =>  =>  list
                    .(+6 [* *])
                %


          cast:
          have  $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    :-  @
                    =>  =>  list
                        .(+6 @)
                    %
                ==

          need  %

          cast:
          have  $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    :-  @
                    =>  =>  list
                        .(+6 @)
                    %
                ==

          need  $:  ?(%7104878 %7237475)
                    ?:  ?=(%"nil" +2)
                      $~
                    :-  [* *]
                    =>  =>  list
                        .(+6 [* *])
                    %
                ==


          cast:
          have  ?:  ?=(%"nil" +2)
                  $~
                :-  @
                =>  =>  list
                    .(+6 @)
                %

          need  ?:  ?=(%"nil" +2)
                  $~
                :-  [* *]
                =>  =>  list
                    .(+6 [* *])
                %


          cast:
          have  :-  @
                =>  =>  list
                    .(+6 @)
                %

          need  :-  [* *]
                =>  =>  list
                    .(+6 [* *])
                %


          cast:
          have  @
          need  [* *]


  cast-fail:
  have  @
  need  [* *]

